cns/requestcontroller/kubecontroller/crdreconciler.go

```diff
@@ -30,30 +34,95 @@ func (r *CrdReconciler) Reconcile(request reconcile.Request) (reconcile.Result,
 			return reconcile.Result{}, client.IgnoreNotFound(err)
 		} else {
 			logger.Errorf("[cns-rc] Error retrieving CRD from cache : %v", err)
+			//requeue
 			return reconcile.Result{}, err
 		}
 	}
 
-	logger.Printf("[cns-rc] CRD object: %v", nodeNetConfig)
+	logger.Printf("[cns-rc] CRD spec: %v", nodeNetConfig.Spec)
+	logger.Printf("[cns-rc] CRD status: %v", nodeNetConfig.Status)
 
-	//TODO: Translate CRD status into NetworkContainer request
-	ncRequest, err := CRDStatusToNCRequest(&nodeNetConfig.Status)
+	ipConfigs, err := CRDStatusToCNS(&nodeNetConfig.Status)
 	if err != nil {
 		logger.Errorf("[cns-rc] Error translating crd status to nc request %v", err)
 		//requeue
 		return reconcile.Result{}, err
 	}
 
-	//TODO: process the nc request on CNS side
-	r.CNSClient.UpdateCNSState(ncRequest)
+	if r.CNSClient.ReadyToIPAM() {
+		if err := r.CNSClient.UpdateCNSState(ipConfigs); err != nil {
+			logger.Errorf("[cns-rc] Error updating CNS state: %v", err)
+			//requeue
+			return reconcile.Result{}, err
+		}
+	} else {
+		if err := r.markIPs(ipConfigs); err != nil {
+			logger.Errorf("[cns-rc] Error marking ips as allocated when readying CNS: %v", err)
+			//requeue
+			return reconcile.Result{}, err
+		}
+
+		if err := r.CNSClient.InitCNSState(ipConfigs); err != nil {
```

Comment:
Is the intention to lazily initialize CNS? Does it ever go back to being uninitialized?

Reply:
Yes the intention is to lazily initialize CNS, the only time it would go back to being uninitialized is on reboot.